Mods/SML/Source/SML/Private/Toolkit/KismetBytecodeDisassembler.cpp

```diff
@@ -3,9 +3,12 @@
 #include "Toolkit/PropertyTypeHandler.h"
 
 TSharedPtr<FJsonObject> FSMLKismetBytecodeDisassembler::SerializeExpression(int32& ScriptIndex) {
-	EExprToken Opcode = (EExprToken) ReadByte(ScriptIndex);
+	int32 StartingScriptIndex = ScriptIndex;
+	EExprToken Opcode = (EExprToken)ReadByte(ScriptIndex);
 	TSharedPtr<FJsonObject> Result = MakeShareable(new FJsonObject());
-	
+
+	Result->SetNumberField(TEXT("Opcode"), Opcode);
+	Result->SetNumberField(TEXT("StatementIndex"), StartingScriptIndex);
```

Comment:
Why is this added here? `StatementIndex` doesn't make sense for inner expressions of a statement, and `Opcode` seems to be redundant given that `Inst` exists.

Reply:
I need the opcode to switch on so I can update jump targets correctly in BlueprintHookManager.  The Inst node is just a short text description of the opcode that I think Arch decided on when he originally wrote the code.  The opcode is the specific instruction code and will always mean the same thing even if the text description changes.

StatementIndex - if the comment is about the name, I agree.  ScriptIndex might make more sense, so I changed it to that but am open to any other suggestions. But having it on every node has been helpful for me when trying to understand the layout of the script and SerializeExpression actually does statements, too, which mean the statement index was only being included in the SerializeFunction code path when it could be relevant on others.  BlueprintHookManager also now assumes this info is on every node, so it reduces special cases.

Reply:
Here is something from https://github.com/Th3Fanbus/Th3ScriptViewer that might help me explain myself:

![ExecuteUbergraph_BP_WAT1_1126 gv](https://github.com/user-attachments/assets/f0ed1aec-9c76-4475-929e-cdee379e2140)

`index` of each "block" is `StatementIndex`. But notice there's inner expressions that have `inst`, but don't have `StatementIndex`.

If you want to keep the index for debugging, I would recommend using a different name, e.g. `EnclosingStatementIndex`.

Reply:
Really, it's index in the script array where the Opcode resides, whether that's for a statement or expression or a subexpression.  And saying it like that makes me think OpcodeIndex is probably the right name.  Here's a quick snapshot to illustrate.

	{
		"Opcode": 7,
		"OpcodeIndex": 869,
		"Inst": "JumpIfNot",
		"Offset": 1634,
		"Condition":
		{
			"Opcode": 0,
			"OpcodeIndex": 874,
			"Inst": "LocalVariable",
			"VariableType":
			{
				"PinCategory": "bool",
				"PinSubCategory": "bool"
			},
			"VariableName": "CallFunc_GetGenericClass_HasGenericClass",
			"ExpressionSizeInBytes": 9
		},
		"ExpressionSizeInBytes": 14
	},

Note that ExpressionSizeInBytes has a similar naming problem and I only chose "ExpressionSize" because the function is named SerializeExpression even though it serializes more than that.  What if I also rename ExpressionSizeInBytes to OperationSizeInBytes?  Or maybe just SizeInBytes?

Reply:
The names for "Inst" are the enumerators' names without the `EX_` prefix. Those have to remain in sync with the enum, which comes from engine code. I don't mind adding the numeric opcode value, though.

In general, I would be careful about renaming JSON fields because other tools use the same names, e.g. FModel, Th3ScriptViewer, the Asset Toolkit... It's true some names are confusing, but I'd say it would be more confusing to have different names for the same thing.

About statements and expressions: I just saw that you're checking all expressions to see if there's a jump offset to modify, but I don't think inner expressions use those. So I feel that there shouldn't be any need to have `StatementIndex` on non-statement expressions. If you find a case where this would happen, could you please point it out?



I decided to explain a bit more now that I am awake: The idea is that the minimum unit of execution is a statement (so it has a statement index). But a statement is composed of one or more expressions. Those expressions cannot be executed on their own, only as part of the enclosing statement. In the JSON, the "root" expression of a statement is embedded within the statement itself; `StatementIndex` is what distinguishes statements from expressions.

For example, in my picture earlier, the first block is a statement:

 - the root is a `EX_LetBool` expression
   - `var` (the destination variable) is a `EX_LocalVariable` expression
   - `expr` (the value to assign) is a `EX_CallMath` (call without context) expression
      - `params` is an array with an `EX_InstanceVariable` expression

(Note: I abbreviated "expression with a `EX_Foo` instruction" to "`EX_Foo` expression")

There's 4 expressions in there, and I've seen statements containing a lot more expressions.

Reply:
Ok, then I think the best compromise is to restore StatementIndex as it was and not include it on expressions, but I can also include OpcodeIndex on every node - it will be redundant for statements but also clarify where expressions are.

Do you have any thoughts about the ExpressionSizeInBytes name?

Reply:
@Th3Fanbus For some incomprehensible reason, GitHub had this unresolved conversation collapsed. Just wanted to be sure you saw my last message and think that's OK.

Reply:
Ack on first paragraph, I wouldn't rename `ExpressionSizeInBytes` if that was the original name.

Reply:
There was no sizeInBytes property there before - I added it and called it ExpressionSizeInBytes because the function was called Serialize**Expression**, even though neither was totally accurate.  The intra-PR rename was just to make it a bit more accurate.

Reply:
Okay, if you added it, feel free to choose any name that makes sense and isn't used for anything else.